### ISSUE TITLE
OpenSSL 1.1.1 update in various repositories (apache2, nginx, php)

### ISSUE THREAD

**Issue by Ayesh** [oerdnj/deb.sury.org#951]
Hi there,
Thanks for your continuous efforts in bringing PHP and Apache builds. 

I just wanted to know if there are any plans to release Apache/PHP builds with OpenSSL 1.1.1, which was released a couple weeks back. It adds TLS 1.3, which I'm eager to try out and I guess it's in the interests of many others too. 

Thank you.

**Comment by ahrasis** [comment]
I am also eager to try this out as well but mine is for Nginx which is currently using 1.1.0h of your ubuntu version. Hopefully, this can be updated to 1.1.1 soon.

Thank you.

**Comment by Ayesh** [comment]
As for TLS 1.3, it doesn't look like Apache 2.4.35 at least, does not add TLSv1.3 to the supported versions list, so you cannot actually enable it. 

If you are interested in testing TLS 1.3 as a standalone server, `openssl s_server` will help you run one. 
For Apache integration, you will have to compile Apache 2.5 with OpenSSL 1.1.1 at the time of writing.

**Comment by oerdnj** [comment]
I would rather let the dust settle before rushing to change the default OpenSSL version.  What about we revisit this after month or two?

**Comment by Ayesh** [comment]
Sound good, @oerdnj . Thank you so much. Apache server itself seems to wait a bit before backporting TLSv1.3 support from the 2.5 branch too, so this kind of goes along with it.

**Comment by ahrasis** [comment]
Yes, we can wait. In any event, I already compiled and installed my own deb for openssl 1.1.1. Once yours are ready, I'll change to it.

**Comment by Ayesh** [comment]
@ahrasis could you get TLSv1.3 to work with Apache 2.4.x releases?

**Comment by Ayesh** [comment]
Apache 2.4.37 is released and now supports OpenSSL 1.1.1 and has TLSv1.3 supported backported from trunk.

**Comment by Ayesh** [comment]
@oerdnj perhaps it's a good time to build 2.4.37 with OpenSSL 1.1.1. what do you think?

**Comment by swiffer** [comment]
excited to try that out 👍

**Comment by Ayesh** [comment]
2.4.37 is not _officially_ released (it was in voting phrase when I submitted my last commit). Really looking forward for this. Thanks.

**Comment by derritter88** [comment]
Ahoj Ondrej! I would also love to have the new Apache version compiled for Ubuntu 16.04. Cheers!

**Comment by ahrasis** [comment]
I hope the same is built for nginx too. However, I noticed the title has just been changed to apache2, so should we opened a new issue for nginx or that would not be necessary?

**Comment by Ayesh** [comment]
Thank you so much @oerdnj, it's working perfectly! Totally grateful for all these. 

@ahrasis - If you are trying to get nginx to pull TLS 1.3, nginx 1.15 frm main nginx repo (and even Ubuntu cosmic sources) has it. For OpenSSL 1.1.1, you can add the apache-qa repo. 
I wrote about Apache (compile yourself) and nginx (install from repo) here: add-apt-repository ppa:ondrej/apache2-qa

**Comment by ahrasis** [comment]
I am still using nginx 1.14 but with self-compiled OpenSSL 1.1.1. So far there is no problem faced and it is good as I can already enabled and used TLSv1.3.
```
nginx version: nginx/1.14.0 (Ubuntu)
built with OpenSSL 1.1.0g  2 Nov 2017 (running with OpenSSL 1.1.1  11 Sep 2018)
```
I hope to see and used a proper one from here soon.

**Comment by oerdnj** [comment]
For nginx mainline try adding `ppa:ondrej/nginx-qa` on top of `ppa:ondrej/nginx-mainline`.

**Comment by Ayesh** [comment]
I see Apache 2.4.37 and openssl 1.1.1 now in the main Apache repo. **Thank you so much!**. For future Googles, I have written a new post on how to enable TLS 1.3 in Apache: https://ayesh.me/TLSv1.3-Apache

**Comment by swiffer** [comment]
hey @Ayesh ,

read your article.
i think configuring ciphersuite works by calling

```
SSLCipherSuite TLSv1.3  TLS_AES_256_GCM_SHA384:TLS_CHACHA20_POLY1305_SHA256
```

see: https://httpd.apache.org/docs/2.4/mod/mod_ssl.html#sslciphersuite
at least for me this works fine ;-)

**Comment by Ayesh** [comment]
Thanks @swiffer - I updated the post 👍.

**Comment by ahrasis** [comment]
> 
> 
> For nginx mainline try adding `ppa:ondrej/nginx-qa` on top of `ppa:ondrej/nginx-mainline`.

Since I was using default Ubuntu 18.04 Nginx 1.14, I have to remove it first before reinstall nginx. Now it display:
```
nginx version: nginx/1.15.5
built with OpenSSL 1.1.0h  27 Mar 2018 (running with OpenSSL 1.1.1  11 Sep 2018)
```
TLSv1.3 works on all websites after I add it to their nginx conf / vhost files ssl_protocols with TLS-CHACHA20-POLY1305-SHA256:TLS-AES-256-GCM-SHA384:TLS-AES-128-GCM-SHA256 as its ssl_ciphers.

**Comment by oerdnj** [comment]
OpenSSL 1.1.1 has been uploaded to `apache2`, `php`, `nginx` and `nginx-stable` repositories.

**Comment by yura3d** [comment]
What is the version of OpenSSL used to build nginx? Cause since 1.15.5 and 1.14.2 it's not enough for using TLS 1.3 to have OpenSSL 1.1.1 installed, nginx must be built with it. I got proof here as an answer on my question from one of nginx contributors: https://trac.nginx.org/nginx/ticket/1654#comment:9

**Comment by ahrasis** [comment]
Check your ciphers. It works well for most of us.

**Comment by yura3d** [comment]
Thanks, of course, I checked ciphers twice. And only after rebuild nginx from sources manually with OpenSSL 1.1.1 I got TLS 1.3 support (with the same ciphers and other config).

If you installed nginx from @oerdnj repo, can you provide your "nginx -V" ouput please?

**Comment by ahrasis** [comment]
My post was just before ondrej, above yours.

**Comment by kevin-olbrich** [comment]
I had the same problem because I was using "nginx" from mainline repo at nginx.org.
While it has TLS 1.3 support, it was not compiled against a capable OpenSSL.
After switching to "nginx-full" which is superseded by sury.org repo, TLS 1.3 works perfectly fine.

**Comment by kevin-olbrich** [comment]
Maybe I missed it but the docs tell me, that nginx supports TLS 1.3 when a capable version of OpenSSL is installed.
This is not the case. It needs to be compiled against it to work.
In the past, patching OpenSSL was sufficient to e.g. protect other applications. This is why I thought, replacing the lib will make TLS 1.3 work.
This might be a similar case for @yura3d as he also mentions, after compiling from sources, it worked.

**Comment by oerdnj** [comment]
I will be updating the nginx both stable and mainline this week, but I am pretty sure the nginx in my repositories were compiled with OpenSSL 1.1.1.

**Comment by yura3d** [comment]
@ahrasis, you are using 1.15.5 built with OpenSSL 1.1.0h (but working with OpenSSL 1.1.1) and everything is fine in this case. The problem starts from 1.15.6 because of this patch: https://trac.nginx.org/nginx/ticket/1654#comment:9
And if you will update to 1.15.6 built with the same (1.1.0h) OpenSSL version, TLS 1.3 support will break, even despite of installed OpenSSL 1.1.1 in your system.

> Maybe I missed it but the docs tell me, that nginx supports TLS 1.3 when a capable version of OpenSSL is installed.
This is not the case. It needs to be compiled against it to work.

@kevin-olbrich yes, you're right. After the patch (which I refered above) [the section for **ssl_protocols**](http://nginx.org/ru/docs/http/ngx_http_ssl_module.html#ssl_protocols) in the nginx docs contains incorrect notification about OpenSSL.

**Comment by todeveni** [comment]
> I will be updating the nginx both stable and mainline this week, but I am pretty sure the nginx in my repositories were compiled with OpenSSL 1.1.1.

```
nginx:
  Installed: 1.14.1-1+0~20181114214155.2+stretch~1.gbp5630c9

nginx version: nginx/1.14.1
built with OpenSSL 1.1.1  11 Sep 2018
```

**Comment by ahrasis** [comment]
```
nginx version: nginx/1.15.6
built with OpenSSL 1.1.1  11 Sep 2018
```
Still works great with TLS v1.3.

**Comment by yura3d** [comment]
> Still works great with TLS v1.3.

Thanks, that was my question: which OpenSSL version is used for building current versions (1.15.6 and 1.14.1). Now I see this is 1.1.1.

**Comment by ahrasis** [comment]
You said you was already using OpenSSL 1.1.1 but eventhough you set your nginx to use TLS v1.3, it was not working for you. The version of nginx used by us is from Ondrej Sury repos which was compiled using OpenSSL 1.1.1. If you was using nginx from other sources, that may be was the cause of your TLS v1.3 failure.
